arch/djgpp/event.c

```diff
@@ -0,0 +1,583 @@
+    case 0xA5: return 0x0F; // Alt-Tab
+    case 0xA6: return 0x1C; // Alt-KP-Enter
+    default: return key & 0x7F;
+  }
+}
+
+static void poll_keyboard_bios(void)
+{
+  unsigned short res;
+  Uint8 scancode;
+  Uint16 unicode;
+
+  while(extbioskey(1))
+  {
+    res = extbioskey(0);
+    scancode = convert_bios_xt(res >> 8);
+    unicode = res & 0xFF;
+    kbd_unicode[scancode] = unicode;
+    if(kbd_statmap[scancode] == KBD_RELEASED)
+      kbd_statmap[scancode] = KBD_PRESSING;
+  }
+}
+
+static Uint16 convert_ext_unicode(Uint8 key)
+{
+  poll_keyboard_bios();
+  return kbd_unicode[key & 0x7F];
+}
+
+static int get_keystat(int key)
+{
+  return kbd_statmap[key & 0x7F];
+}
+
+static void set_keystat(int key, int stat)
+{
+  kbd_statmap[key & 0x7F] = stat;
+}
+
+static boolean non_bios_key(Uint8 key)
+{
+  switch(key)
+  {
+    case 0x1D: // Right Ctrl
+    case 0x9D: // Left Ctrl
+    case 0x38: // Right Alt
+    case 0xB8: // Left Alt
+    case 0x2A: // Left Shift
+    case 0x36: // Right Shift
+    case 0xDB: // Left Super (Windows)
+    case 0xDC: // Right Super (Windows)
+    case 0xDD: // Menu
+      return true;
+    // BIOS can't return keycodes >= 0x54 due to Alt/Ctrl keycode modification
+    // shenanigans, except for F11/F12 which are given alternate keycodes
+    case 0x57:
+    case 0x58:
+      return false;
+    default:
+      if((key & 0x7F) >= 0x54)
+        return true;
+      else
+        return false;
+  }
+}
+
+static boolean process_keypress(int key)
+{
+  struct buffered_status *status = store_status();
+  enum keycode ikey = convert_ext_internal(key);
+  Uint16 unicode = convert_ext_unicode(key);
+
+  if((get_keystat(key) != KBD_PRESSING) && !non_bios_key(key))
+    return false;
+  set_keystat(key, KBD_PRESSED);
+
+  if(!ikey)
+  {
+    if(unicode)
+      ikey = IKEY_UNICODE;
+    else
+      return false;
+  }
+
+  if(status->keymap[ikey])
+    return false;
+
+  if((ikey == IKEY_CAPSLOCK) || (ikey == IKEY_NUMLOCK))
+    update_lock_status(status);
+
+  if((ikey == IKEY_RETURN) &&
+   get_alt_status(keycode_internal) &&
+   get_ctrl_status(keycode_internal))
+  {
+    toggle_fullscreen();
+    return true;
+  }
+
+#ifdef CONFIG_ENABLE_SCREENSHOTS
+  if(ikey == IKEY_F12)
+  {
+    dump_screen();
+    return true;
+  }
+#endif
+
+  if(status->key_repeat &&
+   (status->key_repeat != IKEY_LSHIFT) &&
+   (status->key_repeat != IKEY_RSHIFT) &&
+   (status->key_repeat != IKEY_LALT) &&
+   (status->key_repeat != IKEY_RALT) &&
+   (status->key_repeat != IKEY_LCTRL) &&
+   (status->key_repeat != IKEY_RCTRL))
+  {
+    // Stack current repeat key if it isn't shift, alt, or ctrl
+    if(input.repeat_stack_pointer != KEY_REPEAT_STACK_SIZE)
+    {
+      input.key_repeat_stack[input.repeat_stack_pointer] =
+       status->key_repeat;
+      input.unicode_repeat_stack[input.repeat_stack_pointer] =
+       status->unicode_repeat;
+      input.repeat_stack_pointer++;
+    }
+  }
+
+  key_press(status, ikey, unicode);
```

Comment:
Whoops, missed one. This needs to be changed to
```
  key_press(status, ikey);
  key_press_unicode(status, unicode, true);
```
to build with the lastest master.